nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -1394,6 +1450,210 @@ impl<'a> Planner<'a> {
+                    );
+                } else {
+                    // It would be very strange for the current value to be
+                    // greater than the new value. That would indicate something
+                    // like a row being removed from the target release
+                    // generation table -- one of the invariants of the target
+                    // release generation is that it only moves forward.
+                    //
+                    // In this case we bail out of planning entirely.
+                    return Err(
+                        Error::TargetReleaseMinimumGenerationRollback {
+                            current,
+                            new,
+                        },
+                    );
+                }
+                self.blueprint
+                    .set_target_release_minimum_generation(current, new)
+                    .expect("current value passed in => can't fail");
+            }
+        }
+
+        // Now we need to determine whether to also perform other actions like
+        // updating or adding zones. We have to be careful here:
+        //
+        // * We may have moved existing zones with an Artifact source to using
+        //   the install dataset via the BpSetOverride action, but we don't want
+        //   to use the install dataset on sleds that weren't MUPdated (because
+        //   the install dataset might be ancient).
+        //
+        // * While any overrides are in place according to inventory, we wait
+        //   for the system to recover and don't start new zones on *any* sleds,
+        //   or perform any further updates.
+        //
+        // This condition is level-triggered on the following conditions:
+        //
+        // 1. If the planning input's target release generation is less than the
+        //    minimum generation set in the blueprint, the operator hasn't set a
+        //    new generation in the blueprint -- we should wait to decide what
+        //    to do until the operator provides an indication.
+        //
+        // 2. If any sleds have a mupdate override set in the blueprint, then
+        //    we're still recovering from a MUPdate. If that is the case, we
+        //    don't want to add zones on *any* sled.
+        //
+        //    This might seem overly conservative (why block zone additions on
+        //    *all* sleds if *any* are currently recovering from a MUPdate?),
+        //    but is probably correct for the medium term: we want to minimize
+        //    the number of different versions of services running at any time.
+        //
+        //    There's some potential to relax this in the future (e.g. by
+        //    matching up the zone manifest with the target release to compute
+        //    the number of versions running at a given time), but that's a
+        //    non-trivial optimization that we should probably defer until we
+        //    see its necessity.
```

Comment:
I'm not proposing we put this in this PR, but just want to confirm: I _think_ the thing we've been talking about where the planner shouldn't proceed with an upgrade until all zones have been converted back to non-`InstallDataset`-sources could be a third condition here, right? In particular, step 2 of this comment uses the phrase "recovering from a MUPdate", and any zone still configured to run out of an install dataset means we're still in that state.

Do you think there's more involved (other than testing ofc) than adding a third check at this point?